cmd/kubeadm/app/util/config/masterconfig_test.go

```diff
@@ -14,5 +14,124 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// TODO: write unit tests for the functions in this package
 package config
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/scheme"
+	"k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1alpha1"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
+)
+
+const (
+	v1alpha1YAML                = "testdata/conversion/v1alpha1.yaml"
+	v1alpha1WithoutTypeMetaYAML = "testdata/conversion/v1alpha1_without_TypeMeta.yaml"
+	internalYAML                = "testdata/conversion/internal.yaml"
+	incompleteYAML              = "testdata/defaulting/incomplete.yaml"
+	defaultedYAML               = "testdata/defaulting/defaulted.yaml"
+	invalidYAML                 = "testdata/validation/invalid.yaml"
+)
+
+// TestConversion tests reading one file, loading it using ConfigFileAndDefaultsToInternalConfig that all of kubeadm is using for unmarshal of our API types,
+// and then marshals the internal object to the expected groupVersion
+func TestConversion(t *testing.T) {
+	var tests = []struct {
+		in, out      string
+		groupVersion schema.GroupVersion
+	}{
+		{ // v1alpha1 (faulty) -> internal
+			in:           v1alpha1WithoutTypeMetaYAML,
+			out:          internalYAML,
+			groupVersion: kubeadm.SchemeGroupVersion,
+		},
+		{ // v1alpha1 -> internal
+			in:           v1alpha1YAML,
+			out:          internalYAML,
+			groupVersion: kubeadm.SchemeGroupVersion,
+		},
+		{ // v1alpha1 (faulty) -> internal -> v1alpha1
+			in:           v1alpha1WithoutTypeMetaYAML,
+			out:          v1alpha1YAML,
+			groupVersion: v1alpha1.SchemeGroupVersion,
+		},
+	}
+
+	for _, rt := range tests {
+		err := testRoundTrip(rt.in, rt.out, rt.groupVersion)
+		if err != nil {
+			t.Errorf("%v", err)
+		}
+	}
+}
+
+func testRoundTrip(in, out string, gv schema.GroupVersion) error {
+	expected, err := ioutil.ReadFile(out)
+	if err != nil {
+		return fmt.Errorf("couldn't read test data: %v", err)
+	}
+
+	internalcfg, err := ConfigFileAndDefaultsToInternalConfig(in, &v1alpha1.MasterConfiguration{})
+	if err != nil {
+		return fmt.Errorf("couldn't unmarshal test data: %v", err)
+	}
+
+	actual, err := kubeadmutil.MarshalToYamlForCodecs(internalcfg, gv, scheme.Codecs)
+	if err != nil {
+		return fmt.Errorf("couldn't marshal internal object: %v", err)
+	}
+
+	if !bytes.Equal(expected, actual) {
```

Comment:
maybe use something like https://godoc.org/github.com/kr/pretty#Diff so the diffs are easier to see?

Reply:
An other time, don't think I have time to make that change currently. Please file a generic issue, as we have this in many unit tests that do this